src/react/reconcilation.js

```diff
@@ -510,18 +517,20 @@ export class Reconciler {
 
         this._findReactComponentTrees(propsValue, evaluatedChildNode, "NORMAL_FUNCTIONS");
         if (renderProp instanceof ECMAScriptSourceFunctionValue) {
-          if (this.componentTreeConfig.firstRenderOnly) {
-            if (typeValue instanceof ObjectValue || typeValue instanceof AbstractObjectValue) {
-              // make sure this context is in our tree
-              if (this._hasReferenceForContextNode(typeValue)) {
-                let valueProp = Get(this.realm, typeValue, "currentValue");
+          if (typeValue instanceof ObjectValue || typeValue instanceof AbstractObjectValue) {
+            // make sure this context is in our tree
+            if (this._hasReferenceForContextNode(typeValue)) {
+              let valueProp = Get(this.realm, typeValue, "currentValue");
+              // if the value is abstract, we need to keep the render prop as unless
+              // we are in firstRenderOnly mode, where we can just inline the abstract value
+              if (!(valueProp instanceof AbstractValue) || this.componentTreeConfig.firstRenderOnly) {
                 let result = getValueFromFunctionCall(this.realm, renderProp, this.realm.intrinsics.undefined, [
                   valueProp,
                 ]);
                 this.statistics.inlinedComponents++;
                 this.statistics.componentsEvaluated++;
                 evaluatedChildNode.status = "INLINED";
-                return result;
+                return this._resolveDeeply(componentType, result, context, branchStatus, evaluatedNode);
```

Comment:
What does this fix? It doesn't seem to affect the test result.

Reply:
It further inlines code, if you remove this line, one of the newer tests fail for me (as the contents of the render props doesn't get inlined).

Reply:
Is this using nested optimized functions under the hood? Shouldn't we also disable this when `reactOptimizeNestedFunctions` is `false`? I'm worried those are too buggy to rely on.

Reply:
Ah, no, this is not trying to optimize nested optimized functions (which are caught later by `reactOptimizeNestedFunctions` flag anyway, so that's handled). This is when we don't need to process the render prop as an optimized nested function and can just inline the contents.

Reply:
What's the difference between these two cases? Both are functions inside functions. Can you explain more about how this callback differs from e.g. `map` callback?

Reply:
No, this is not a function, this is inlining the result from calling the render prop on line 527.

Reply:
Ahh I see. You don't need to leave function itself in the output. Got it.